README.md
<!--[![Build Status](https://travis-ci.org/castlabs/dashas.png?branch=master)](https://travis-ci.org/castlabs/dashas)-->

dash.as
=======

Dash.as plays [MPEG-DASH](http://dashif.org/) video on any device supporting Adobe Flash. It is written in Adobe ActionScript and was designed from the ground-up to be lightweight with performance in mind. 

MPEG-DASH (Dynamic Adaptive Streaming over HTTP) became an [International Standard](http://standards.iso.org/ittf/PubliclyAvailableStandards/c057623_ISO_IEC_23009-1_2012.zip) in 2011 and is a technique designed specifically for streaming high-quality multimedia over an HTTP server. It works by fragmenting content into a series of small HTTP-based file segments where each contains block of playback time. Content is made available at a variety of different bitrates to account for variations in a connection's bandwidth quality. Upon each download of a segment, the highest bitrate the connection can handle in time for seamless playback is selected. This allows MPEG-DASH to seamlessly adapt to changing network conditions at the highest quality possible for a seamless user experience.

This project is an extension for [Open Source Media Framework](http://osmf.org/) and released under the [Mozilla Public License 2.0](http://www.mozilla.org/MPL/2.0/).

At the same time there is profersional edition of dash.as which integrates nicely with the [Adobe Access DRM](http://www.adobe.com/products/adobe-access.html). For more information please contact with sales department: "sales (at) castlabs (dot) com".

Please visit [CHANGELOG](https://github.com/castlabs/dashas/blob/master/CHANGELOG.md) for changes history.

For information on a digital rights management (DRM) enabled version, please visit the  [castLabs](http://castlabs.com/products/dashas/) website.

## Demo

For more information visit [examples page](http://dashas.castlabs.com/demo/try.html).

## Features

* An OSMF extension;
* Basic support for manifests with a list [1];
* Basic support for manifests with a template [1];
* Basic support for manifests with a base URL [1][2];
* Basic support for manifests with a time line;
* Audio-video playback [3];
* Adaptive bitrate streaming;
* Live streaming;
* User interface:
	* Play/Pause;
	* Seeking;
	* Fullscreen mode;
	* Duration;
	* Loaded bytes indicator;
	* Buffering indicator;

[1] A server has to have [crossdomain.xml](https://github.com/castlabs/dashas/blob/master/utils/crossdomain.xml) file in the root context.  
[2] A server has to have custom [.htaccess](https://github.com/castlabs/dashas/blob/master/utils/.htaccess) file in the root context.  
[3] Adobe Flash Player supports only [H.264 (MPEG-4 Part 10)  and HE-AAC (MPEG-4 Part 3) codecs](http://helpx.adobe.com/flash/kb/supported-codecs-flash-player.html).

For more information about which attributes and tags are required in a manifest please visit [unit tests](https://github.com/castlabs/dashas/blob/master/src/test/actionscript/com/castlabs/dash/ManifestHandlerTest.as).

## Quick start

### Input files

Create media input files in different bitrates.

#### Prerequisites

* ffmpeg is installed with libfdk_aac and libx264 library (tested with version 2.1.3);


#### Instructions

1. Download source file:

		$ wget http://mirrorblender.top-ix.org/movies/sintel-1024-surround.mp4

1. Create video file encoded at 250k 436p, H.264 codec:

		$ ffmpeg -i sintel-1024-surround.mp4 -an -b:v 250k -s 1024x436 -vcodec libx264 avc1-sintel-436p-250k.mp4

1. Create video file encoded at 500k 436p, H.264 codec:

		$ ffmpeg -i sintel-1024-surround.mp4 -an -b:v 500k -s 1024x436 -vcodec libx264 avc1-sintel-436p-500k.mp4
		
1. Create video file encoded at 1000k 436p, H.264 codec:

		$ ffmpeg -i sintel-1024-surround.mp4 -an -b:v 1000k -s 1024x436 -vcodec libx264 avc1-sintel-436p-1000k.mp4
		
1. Create audio file encoded at 69k stereo, HE-AAC codec:

		$ ffmpeg -i sintel-1024-surround.mp4 -vn -ac 2 -acodec libfdk_aac -profile:a aac_he -f mp4 mp4a-sintel-69k.mp4
		
		
### Stream
		
Convert media input files into DASH stream files.
		
#### Prerequisites

* MP4Box is installed (tested with version 0.5.1r5152);

#### Instructions

* Use following command to generate DASH stream:
		
		$ MP4Box -dash 3000 -bs-switching no -segment-name "%s_" -url-template -out sintel.mpd avc1-sintel-436p-250k.mp4 avc1-sintel-436p-500k.mp4 avc1-sintel-436p-1000k.mp4 mp4a-sintel-69k.mp4
		
### Page

Finally try dash.as locally.

#### Prerequisites

* apache2 is installed;

#### Instructions

1. Create a virtual host for localhost;

1. Copy DASH stream files into a document root;

1. Copy [crossdomain.xml](https://github.com/castlabs/dashas/blob/master/utils/crossdomain.xml) file into a document root (for more information visit [crossdomain page](https://github.com/castlabs/dashas/wiki/crossdomain));

1. Copy [.htaccess](https://github.com/castlabs/dashas/blob/master/utils/.htaccess) file into a document root (for more information visit [htaccess page](https://github.com/castlabs/dashas/wiki/htaccess));

1. Copy contents of the [demo](https://github.com/castlabs/dashas/tree/master/site/demo/) directory into a document root;

1. Create `index.html` file in a document root (type absolute URL to a `sintel.mpd` file):

		<html>
		<head>
		<script type="text/javascript" src="/swfobject/swfobject.js"></script>
		<script type="text/javascript">
		    var flashvars = {};
		
		    // absolut URL to sintel.mpd file
		    flashvars.src = encodeURIComponent("<absolute URL to sintel.mpd file>");
		
		    // absolut URL to dashas.swf file
		    flashvars.plugin_DashPlugin = encodeURIComponent(location.href + "/production/dashas.swf");
		
		    var params = {};
		    params.allowfullscreen = "true";
		    params.allownetworking = "true";
		    params.wmode = "direct";
		
		    swfobject.embedSWF("/StrobeMediaPlayback.swf", "placeholder", "640", "360", "10.1", "/swfobject/expressInstall.swf", flashvars, params, {});
		</script>
		</head>
		
		<body>
		<div id="placeholder">
		    <p><span>Please install <a href="http://get.adobe.com/flashplayer/">Adobe Flash Player</a></span></p>
		</div>
		</body>
		</html>
	
1. Go to `http://localhost/index.html`;

---

**TIP:** The flash is very aggressive in caching. If you notice that player doesn't respond to your server side changes, please clean browser's cache (i.e. cached images and files). 
	

## Development

At the moment project supports only Gradle 1.11. Please read [how to install Gradle 1.11 on Linux](https://github.com/castlabs/dashas/wiki/linux).

### Build

#### Prerequisites

* gradle is installed (tested with version 1.11);

#### Instructions

* Build debug SWF package:

		$ cd <project_workspace>
		$ gradle clean compile

* Build production SWF package:

		$ cd <project_workspace>
		$ gradle -Pprofile=production clean compile

... do you prefer developing in an IDE? Read how to [import project into the IntelliJ IDEA](https://github.com/castlabs/dashas/wiki/IntelliJ-IDEA).

### Test

#### Prerequisites

* gradle is installed (tested with version 1.11);

#### Instructions

1. Download and extract Flash Player:
    * [Windows](http://download.macromedia.com/pub/flashplayer/updaters/10/flashplayer_10_sa_debug.exe)
    * [Linux](http://download.macromedia.com/pub/flashplayer/updaters/10/flashplayer_10_sa_debug.tar.gz)
    * [OS X](http://download.macromedia.com/pub/flashplayer/updaters/10/flashplayer_10_sa_debug.app.zip)

1. Define global variable:
    
        $ export FLASH_PLAYER_EXE=Flash\ Player\ Debugger.app/Contents/MacOS/Flash\ Player\ Debugger
        
1. Run test suites:

        $ cd <project_workspace>
        $ gradle test



CHANGELOG.md
## 1.0.3 (2014-12-29)

Bugfixes:

 - Fix rounding issue (@tomasz-oponowicz);
 - Fragment files not being loaded (@tomasz-oponowicz, #31);

## 1.0.2 (2014-09-05)

Bugfixes:

  - Last seconds of stream will never get played back (@tomasz-oponowicz, #25)
  - Support for Audio Sample Entry Box with sound version 1 and 2 (@tomasz-oponowicz)
  - [draft] MPDs without audio tracks are not loaded (@tomasz-oponowicz, #14)
  - Support live streams defined with SegmentTemplate (@tomasz-oponowicz, #23)
  - Inject context listener to every object (@tomasz-oponowicz, #6)

## 1.0.1 (2014-07-31)

Bugfixes:

  - Building process on Linux randomly fails (@tomasz-oponowicz, #20)


site/demo/js/console.js
if(!window.console){ window.console = {log: function(){} }; }

var realUserBandwidths = [], averageUserBandwidths = [], videoBandwidths = [], audioBandwidths = [];

var chartEmptyData = {
    labels : [],
    datasets : [{
        data : []
    }]
};

var chartOptions = {
    animation : false,
    bezierCurve: false,
    datasetFill: false,
    scaleShowGridLines: false
};

function showOrHideMessage(level, node) {
    if (document.getElementById(level).checked) {
        node.style.display = "block";
    } else {
        node.style.display = "none";
    }
}

function showOrHideMessages(level) {
    var nodes = document.getElementsByClassName(level);
    for (var i = 0; i < nodes.length; i++) {
        showOrHideMessage(level, nodes[i]);
    }
}

function handleEvents(events) {
    for (var i = 0; i < events.length; i++) {
        var event = events[i];

        if (event.id == "log") {
            log(event.level, event.message);
        }

        if (event.id == "appendUserBandwidth") {
            appendUserBandwidth(event.dataset, event.bandwidth);
        }

        if (event.id == "appendMediaBandwidth") {
            appendMediaBandwidth(event.dataset, event.bandwidth);
        }
    }
}

function log(level, message) {
    logToPageConsole(level, message);
    logToBrowserConsole(level, message);
}

function logToPageConsole(level, message) {
    if (level == 'error') {
        console.error(message);
    }
    if (level == 'warn') {
        console.warn(message);
    }
    if (level == 'info' || level == 'debug') {
        console.info(message);
    }
}

function logToBrowserConsole(level, message) {
    var node = document.createElement("p");

    node.className = level;
    node.innerHTML = message;
    showOrHideMessage(level, node);

    document.getElementById("screen").appendChild(node);
}

function appendUserBandwidth(dataset, bandwidth) {
    if (dataset == 'real') {
        realUserBandwidths.push(bandwidth);
    }

    if (dataset == 'average') {
        averageUserBandwidths.push(bandwidth);
    }

    limitTo15Entries(realUserBandwidths);
    limitTo15Entries(averageUserBandwidths);

    var data = {
        labels : buildArrayWithEmptyValues(Math.max(realUserBandwidths.length, averageUserBandwidths.length)),
        datasets : [
            {
                strokeColor : "rgba(220,220,220,1)",
                pointColor : "rgba(220,220,220,1)",
                pointStrokeColor : "#fff",
                data : realUserBandwidths
            },
            {
                pointColor : "rgba(151,187,205,1)",
                pointStrokeColor : "#fff",
                strokeColor : "rgba(151,187,205,1)",
                data : averageUserBandwidths
            }
        ]
    };

    var context = document.getElementById("userBandwidthChart").getContext("2d");
    new Chart(context).Line(data, chartOptions);
}

function buildArrayWithEmptyValues(length) {
    var array = [];

    for (var i = 0; i < length; i++) {
        array[i] = "";
    }

    return array;
}

function limitTo15Entries(array) {
    while (array.length > 15) {
        array.shift();
    }
}

function appendMediaBandwidth(dataset, bandwidth) {
    if (dataset == 'video') {
        videoBandwidths.push(bandwidth);
    }

    if (dataset == 'audio') {
        audioBandwidths.push(bandwidth);
    }

    limitTo15Entries(videoBandwidths);
    limitTo15Entries(audioBandwidths);

    var data = {
        labels : buildArrayWithEmptyValues(Math.max(videoBandwidths.length, audioBandwidths.length)),
        datasets : [
            {
                strokeColor : "rgba(151,187,205,1)",
                pointStrokeColor : "#fff",
                pointColor : "rgba(151,187,205,1)",
                data : videoBandwidths
            },
            {
                strokeColor : "rgba(220,220,220,1)",
                pointStrokeColor : "#fff",
                pointColor : "rgba(220,220,220,1)",
                data : audioBandwidths
            }
        ]
    };

    var context = document.getElementById("mediaBandwidthChart").getContext("2d");
    new Chart(context).Line(data, chartOptions);
}

function appendConsole() {
    var node = document.createElement("div");

    node.className = 'console';
    node.innerHTML = '<h4>User\'s Bandwidth</h4><div class="legend"><span class="legend1">average</span><span class="legend2">real</span></div><canvas id="userBandwidthChart" width="640" height="300"></canvas><h4>Media Bandwidth</h4><div class="legend"><span class="legend1">video</span><span class="legend2">audio</span></div><canvas id="mediaBandwidthChart" width="640" height="300"></canvas><h4>Console</h4><div class="log"><div class="buttons"><label for="error"><input id="error" type="checkbox" checked="checked" value=""> Error </label><label for="warn"><input id="warn" type="checkbox" checked="checked" value="">Warn</label><label for="info"><input id="info" type="checkbox" checked="checked" value="">Info</label><label for="debug"><input id="debug" type="checkbox" value="">Debug</label></div><div id="screen"></div></div><h4>Debug</h4><div>Use a debug SWF file: <button id="enableOrDisableDebug" type="button">Enable</button></div>';
    document.body.appendChild(node);
}

function isDebug() {
    return document.cookie.indexOf("debug=true") != -1;
}

function enableDebug() {
    var date = new Date();
    date.setTime(date.getTime() + (365*24*60*60*1000)); // 1 year

    var expires = date.toGMTString();

    document.cookie="debug=true; expires=" + expires;
}

function disabledDebug() {
    document.cookie="debug=; expires=Thu, 01 Jan 1970 00:00:00 GMT";
}

function initEnableOrDisableDebug() {
    var e = document.getElementById("enableOrDisableDebug");

    if (isDebug()) {
        e.textContent = "Disable";

        var n = document.createElement("p");
        n.className = "alert warn";
        n.innerHTML = "You're using a debug SWF file. Dropped frames during playback can occur. <a class='close push-right' href='#'>&times;</a>";

        document.body.insertBefore(n, document.getElementById("placeholder-wrapper"));
    } else {
        e.textContent = "Enable";
    }
}

function initCharts() {
    var context = null;

    context = document.getElementById("userBandwidthChart").getContext("2d");
    new Chart(context).Line(chartEmptyData, chartOptions);

    context = document.getElementById("mediaBandwidthChart").getContext("2d");
    new Chart(context).Line(chartEmptyData, chartOptions);
}

function resetChartsAndLogs() {

    // charts
    realUserBandwidths = [];
    averageUserBandwidths = [];
    videoBandwidths = [];
    audioBandwidths = [];

    appendUserBandwidth(null, null);
    appendMediaBandwidth(null, null);

    // logs
    var node = document.getElementById("screen");
    while (node.firstChild) {
        node.removeChild(node.firstChild);
    }
}

window.addEventListener("load", function() {
    appendConsole();

    initCharts();

    document.getElementById("error").onchange = function() {
        showOrHideMessages("error");
    };
    document.getElementById("warn").onchange = function() {
        showOrHideMessages("warn");
    };
    document.getElementById("info").onchange = function() {
        showOrHideMessages("info");
    };
    document.getElementById("debug").onchange = function() {
        showOrHideMessages("debug");
    };

    initEnableOrDisableDebug();
    document.getElementById("enableOrDisableDebug").onclick = function() {
        if (isDebug()) {
            disabledDebug();
        } else {
            enableDebug();
        }

        location.reload();
    }
}, false);

site/demo/js/try.js
if(!window.console){ window.console = { log: function() {} }; }

var realUserBandwidths = [], averageUserBandwidths = [], videoBandwidths = [], audioBandwidths = [];

var chartEmptyData = {
    labels : [],
    datasets : [{
        data : []
    }]
};

var chartOptions = {
    animation : false,
    bezierCurve: false,
    datasetFill: false,
    scaleShowGridLines: false
};

var playerBridge = null;

function onJavaScriptBridgeCreated(playerId) {
    if (playerBridge == null) {
        playerBridge = document.getElementById(playerId);
    }
}

// call from ActionScript
function handleEvents(events) {
    for (var i = 0; i < events.length; i++) {
        var event = events[i];

        if (event.id == "log") {
            log(event.level, event.message);
        }

        if (event.id == "appendUserBandwidth") {
            appendUserBandwidth(event.dataset, event.bandwidth);
        }

        if (event.id == "appendMediaBandwidth") {
            appendMediaBandwidth(event.dataset, event.bandwidth);
        }
    }
}

function log(level, message) {
    logToPageConsole(level, message);
    logToBrowserConsole(level, message);
}

function logToPageConsole(level, message) {
    if (level == 'error') {
        console.error(message);
    }
    if (level == 'warn') {
        console.warn(message);
    }
    if (level == 'info' || level == 'debug') {
        console.info(message);
    }
}

function logToBrowserConsole(level, message) {
    var node = document.createElement("p");

    node.className = level;
    node.innerHTML = message;
    showOrHideMessage(level, node);

    document.getElementById("screen").appendChild(node);
}

// workaround: firefox can't handle width and height properly
function fixChartSize(chart) {
    chart.width = 640;
    chart.height = 300;
}

function initChart(chart) {
    fixChartSize(chart);
    var context = chart.getContext("2d");
    new Chart(context).Line(chartEmptyData, chartOptions);
}

function appendUserBandwidth(dataset, bandwidth) {
    fixChartSize(document.getElementById("userBandwidthChart"));

    if (dataset == 'real') {
        realUserBandwidths.push(bandwidth);
    }

    if (dataset == 'average') {
        averageUserBandwidths.push(bandwidth);
    }

    limitTo15Entries(realUserBandwidths);
    limitTo15Entries(averageUserBandwidths);

    var data = {
        labels : buildArrayWithEmptyValues(Math.max(realUserBandwidths.length, averageUserBandwidths.length)),
        datasets : [
            {
                strokeColor : "rgba(220,220,220,1)",
                pointColor : "rgba(220,220,220,1)",
                pointStrokeColor : "#fff",
                data : realUserBandwidths
            },
            {
                pointColor : "rgba(151,187,205,1)",
                pointStrokeColor : "#fff",
                strokeColor : "rgba(151,187,205,1)",
                data : averageUserBandwidths
            }
        ]
    };

    var context = document.getElementById("userBandwidthChart").getContext("2d");
    new Chart(context).Line(data, chartOptions);
}

function appendMediaBandwidth(dataset, bandwidth) {
    fixChartSize(document.getElementById("mediaBandwidthChart"));

    if (dataset == 'video') {
        videoBandwidths.push(bandwidth);
    }

    if (dataset == 'audio') {
        audioBandwidths.push(bandwidth);
    }

    limitTo15Entries(videoBandwidths);
    limitTo15Entries(audioBandwidths);

    var data = {
        labels : buildArrayWithEmptyValues(Math.max(videoBandwidths.length, audioBandwidths.length)),
        datasets : [
            {
                strokeColor : "rgba(151,187,205,1)",
                pointStrokeColor : "#fff",
                pointColor : "rgba(151,187,205,1)",
                data : videoBandwidths
            },
            {
                strokeColor : "rgba(220,220,220,1)",
                pointStrokeColor : "#fff",
                pointColor : "rgba(220,220,220,1)",
                data : audioBandwidths
            }
        ]
    };

    var context = document.getElementById("mediaBandwidthChart").getContext("2d");
    new Chart(context).Line(data, chartOptions);
}

function limitTo15Entries(array) {
    while (array.length > 15) {
        array.shift();
    }
}

function buildArrayWithEmptyValues(length) {
    var array = [];

    for (var i = 0; i < length; i++) {
        array[i] = "";
    }

    return array;
}

function isDebug() {
    return document.cookie.indexOf("debug=true") != -1;
}

function enableDebug() {
    var date = new Date();
    date.setTime(date.getTime() + (365*24*60*60*1000)); // 1 year

    var expires = date.toGMTString();

    document.cookie="debug=true; expires=" + expires;
}

function disabledDebug() {
    document.cookie="debug=; expires=Thu, 01 Jan 1970 00:00:00 GMT";
}

function initEnableOrDisableDebug() {
    var e = document.getElementById("enableOrDisableDebug");

    if (isDebug()) {
        e.textContent = "Disable";
        document.getElementById("performance-issue").style.display = "block";
    } else {
        e.textContent = "Enable";
    }
}

function showOrHideMessages(level) {
    var nodes = document.getElementById("screen").getElementsByClassName(level);
    for (var i = 0; i < nodes.length; i++) {
        showOrHideMessage(level, nodes[i]);
    }
}

function showOrHideMessage(level, node) {
    if (document.getElementById(level).checked) {
        node.style.display = "block";
    } else {
        node.style.display = "none";
    }
}

function resetChartsAndLogs() {

    // charts
    realUserBandwidths = [];
    averageUserBandwidths = [];
    videoBandwidths = [];
    audioBandwidths = [];

    appendUserBandwidth(null, null);
    appendMediaBandwidth(null, null);

    // logs
    var node = document.getElementById("screen");
    while (node.firstChild) {
        node.removeChild(node.firstChild);
    }
}

function load() {
    unloadSwf();
    resetChartsAndLogs();
    loadSwf(document.getElementById("url").value);
}

function unloadSwf() {
    swfobject.removeSWF("placeholder");
}

function loadSwf(manifestUrl) {

    // reset bridge to allow initialization after switching do different example
    playerBridge = null;

    document.getElementById("placeholder-wrapper").innerHTML = "<div id='placeholder'><p><span>Please install <a href='http://get.adobe.com/flashplayer/'>Adobe Flash Player</a></span></p></div>";

    var timestamp = new Date().getTime();

    var flashvars = {};
    flashvars.src = encodeURIComponent(manifestUrl);
    flashvars.plugin_WatermarkPlugin = encodeURIComponent(location.href + "/../" + (isDebug() ? "debug" : "production") + "/watermark.swf?t=" + timestamp);
    flashvars.plugin_DashPlugin = encodeURIComponent(location.href + "/../"  + (isDebug() ? "debug" : "production") + "/dashas.swf?t=" + timestamp + "&log=true");
    flashvars.javascriptCallbackFunction = "onJavaScriptBridgeCreated";

    var params = {};
    params.allowfullscreen = "true";
    params.allownetworking = "true";
    params.wmode = "direct";

    swfobject.embedSWF("./StrobeMediaPlayback.swf?=" + timestamp, "placeholder", "640", "360", "10.1", "./swfobject/expressInstall.swf", flashvars, params, {});
}

window.addEventListener("load", function() {
    initChart(document.getElementById("userBandwidthChart"));
    initChart(document.getElementById("mediaBandwidthChart"));

    document.getElementById("error").onchange = function() {
        showOrHideMessages("error");
    };
    document.getElementById("warn").onchange = function() {
        showOrHideMessages("warn");
    };
    document.getElementById("info").onchange = function() {
        showOrHideMessages("info");
    };
    document.getElementById("debug").onchange = function() {
        showOrHideMessages("debug");
    };

    initEnableOrDisableDebug();
    document.getElementById("enableOrDisableDebug").onclick = function() {
        if (isDebug()) {
            disabledDebug();
        } else {
            enableDebug();
        }

        location.reload();
    };

    // hide alerts if press close button
    var elements = document.getElementsByClassName("close");
    for (var i = 0; i < elements.length; i++) {
        elements[i].onclick = function() {
            this.parentNode.style.display = 'none';
            return false;
        };
    }

    // on click load button
    document.getElementById("load").onclick = function() {
        load();
        return false;
    };

    // on change examples selects
    document.getElementById("examples").onchange = function() {
        document.getElementById("url").value = this.value;
        load();

        // select "Examples" option
        this.selectedIndex = 0;
    };

    // load manifest if any
    load();

}, false);